store/bigtable/bigtable.go

```diff
@@ -0,0 +1,471 @@
+			muts[i], n = mutationFromWriteRequest(cwr)
+			//record how long the chunk waited in the queue before we attempted to save to bigtable
+			btblPutWaitDuration.Value(time.Now().Sub(cwr.Timestamp))
+			bytesPerFlush += n
+		}
+		chunkBytesAtSave.Value(bytesPerFlush)
+		success := false
+		attempts := 0
+		for !success {
+			pre := time.Now()
+			ctx, cancel := context.WithTimeout(context.Background(), s.cfg.WriteTimeout)
+			errs, err := s.tbl.ApplyBulk(ctx, rowKeys, muts)
+			cancel()
+			btblPutExecDuration.Value(time.Since(pre))
+			if err != nil {
+				// all chunks in the batch failed to be written.
+				log.Errorf("btStore: unable to apply writes to bigtable. %s", err)
+				chunkSaveFail.Add(len(rowKeys))
+				if (attempts % 20) == 0 {
+					log.Warnf("btStore: Failed to write %d chunks to bigtable. they will be retried. %s", len(rowKeys), err)
+				}
+				sleepTime := 100 * attempts
+				if sleepTime > 2000 {
+					sleepTime = 2000
+				}
+				time.Sleep(time.Duration(sleepTime) * time.Millisecond)
+				attempts++
+			} else if len(errs) > 0 {
+				// only some chunks in the batch failed to get written.
+				failedRowKeys := make([]string, 0)
+				failedMutations := make([]*bigtable.Mutation, 0)
+				retryBuf := make([]*mdata.ChunkWriteRequest, 0)
+				for i, err := range errs {
+					if err != nil {
+						failedRowKeys = append(failedRowKeys, rowKeys[i])
+						failedMutations = append(failedMutations, muts[i])
+						retryBuf = append(retryBuf, buf[i])
+					} else {
+						buf[i].Metric.SyncChunkSaveState(buf[i].Chunk.T0)
+						mdata.SendPersistMessage(buf[i].Key.String(), buf[i].Chunk.T0)
+						log.Debugf("btStore: save complete. %s:%d %v", buf[i].Key, buf[i].Chunk.T0, buf[i].Chunk)
+						chunkSaveOk.Inc()
+					}
+				}
+				rowKeys = failedRowKeys
+				muts = failedMutations
+				buf = retryBuf
+				log.Errorf("btStore: failed to write %d rows. %s", len(failedRowKeys), err)
+				chunkSaveFail.Add(len(failedRowKeys))
+				sleepTime := 100 * attempts
+				if sleepTime > 2000 {
+					sleepTime = 2000
+				}
+				time.Sleep(time.Duration(sleepTime) * time.Millisecond)
+				attempts++
+			} else {
+				success = true
+				chunkSaveOk.Add(len(rowKeys))
+				log.Debugf("btStore: %d chunks saved to bigtable.", len(rowKeys))
+				for _, cwr := range buf {
+					cwr.Metric.SyncChunkSaveState(cwr.Chunk.T0)
+					mdata.SendPersistMessage(cwr.Key.String(), cwr.Chunk.T0)
+					log.Debugf("btStore: save complete. %s:%d %v", cwr.Key.String(), cwr.Chunk.T0, cwr.Chunk)
+				}
+			}
+		}
+		buf = buf[:0]
+	}
+	for {
+		select {
+		case <-timer.C:
+			timer.Reset(time.Second)
+			if len(buf) > 0 {
+				flush()
+			}
+		case cwr := <-queue:
+			buf = append(buf, cwr)
+			if len(buf) >= s.cfg.WriteMaxFlushSize {
+				// make sure the timer hasn't already fired. If it has we read
+				// from the chan and consume the event.
+				if !timer.Stop() {
+					<-timer.C
+				}
+				timer.Reset(time.Second)
+				flush()
+			}
+		case <-s.shutdown:
+			return
```

Comment:
shouldn't we make sure that all pending Add calls are finished writing into queues and that all queues are drained/flushed?

Reply:
Maybe.
a) the cassandraStore just closes the cassandra client connection immediately
b) There is no point in writing to the store when shutting down unless the persist messages are being committed.  So we would need to make sure that the notifier plugin doesnt get closed first.

So it is something we can look at, but i think it is outside the scope of this PR.

Reply:
looking at this and the cassandra code again, there is not the data loss vector i thought there was, as we only send notifier messages (SyncChunkSaveState) after writing to cassandra/bigtable.
you're right that if we closed the notifier, there is no more meed to save chunks, though that sounds a bit backwards : we should probably first finish writing chunks with a reasonable timeout, then close notifier. but i agree, not something to worry about now.